src/Shared/QueryStringEnumerable.cs

```diff
@@ -88,21 +91,20 @@ public ReadOnlyMemory<char> DecodeName()
         public ReadOnlyMemory<char> DecodeValue()
             => Decode(EncodedValue);
 
-        private static unsafe ReadOnlyMemory<char> Decode(ReadOnlyMemory<char> chars)
+        private static ReadOnlyMemory<char> Decode(ReadOnlyMemory<char> chars)
         {
             // If the value is short, it's cheap to check up front if it really needs decoding. If it doesn't,
             // then we can save some allocations.
-            if (chars.Length < 16 && chars.Span.IndexOfAny('%', '+') < 0)
+            ReadOnlySpan<char> source = chars.Span;
+            if (source.Length < 16 && source.IndexOfAny(DecodingChars) < 0)
             {
                 return chars;
             }
-
-#pragma warning disable CS8500 // This takes the address of, gets the size of, or declares a pointer to a managed type
-            ReadOnlySpan<char> span = chars.Span;
-            return Uri.UnescapeDataString(
-                string.Create(span.Length,
-                    (IntPtr)(&span), static (dest, ptr) => ((ReadOnlySpan<char>*)ptr)->Replace(dest, '+', ' '))).AsMemory();
-#pragma warning restore CS8500
+            var buffer = new char[source.Length];
+            source.Replace(buffer, '+', ' ');
+            var success = Uri.TryUnescapeDataString(buffer, buffer, out var unescapedLength);
+            Debug.Assert(success);
+            return buffer.AsMemory(0, unescapedLength);
```

Comment:
1. The difference here to commit [Removing unsafe parts of Decode method in QueryStringEnumerable](https://github.com/dotnet/aspnetcore/pull/56630/commits/988c8680d1e87d1ab7a29da14ce4fc6c3238dd17) is that this implementation keeps a reference to the 'buffer' array that might be larger to the return value with the `AsMemory()` call. On the other hand, this implementation avoids an extra `string` allocation as shown on the perf measurement's *Encoded* line - and as well faster in that scenario.

1. I think `Uri.TryUnescapeDataString` should not return false, despite [this comment](https://source.dot.net/#System.Private.Uri/System/UriHelper.cs,344). The unescape level used by `TryUnescapeDataString` method seems to avoid the 'rare' case. As an alternative the code could branch based on the result of `TryUnescapeDataString`, and in case of a `false` value invoke the `Uri.UnescapeDataString` method as a fallback. 
Or as another alternative, go with the implementation from the previous commit.

Reply:
I agree, TryUnescapeDataString shouldn't be able to return false in our case.
https://source.dot.net/#System.Private.Uri/System/UriExt.cs,648